File: tests/Tests.FSharp/RendezvousHash.CrossVerify.Tests.fs
module Zeta.Tests.RendezvousHashCrossVerifyTests

open System.Globalization
open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

// ═══════════════════════════════════════════════════════════════════
// Rendezvous (HRW) consistent-hash cross-language agreement — the F# oracle uses the REAL canonical
// RendezvousHash.Create / Pick and replays the shared seed
// (src/Core.TypeScript/consistent-hash/golden-vectors.json) that the C#/TS/Rust oracles also verify.
// Pure wrapping uint64 (the SplitMix64 score). F# seeds are private, so the F# leg verifies Pick (the
// observable) — agreement transitively confirms the seed(i)=mix(i) formula matches; C#/TS/Rust verify
// the seeds array directly. Jump is excluded (f64, out of the proof lineage).
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private u (s: string) : uint64 = System.UInt64.Parse(s, CultureInfo.InvariantCulture)

[<Fact>]
let ``F# RendezvousHash.Pick agrees with the shared golden seed`` () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "consistent-hash", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    use doc = JsonDocument.Parse(File.ReadAllText path)
    let picks = doc.RootElement.GetProperty("pick").EnumerateArray() |> Seq.toArray

    for v in picks do
        let n = (v.GetProperty "buckets").GetInt32()
        let key = u ((v.GetProperty "key").GetString())
        let h = RendezvousHash.Create n
        Assert.Equal<int>((v.GetProperty "result").GetInt32(), h.Pick key)

File: tests/Tests.CSharp/RendezvousHashCrossVerifyTests.cs
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

/// <summary>
/// Rendezvous consistent-hash cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/consistent-hash/golden-vectors.json</c>) that the F#/TS/Rust oracles also
/// verify. Pure wrapping uint64 (the SplitMix64 score); uint64 carried as decimal strings.
/// </summary>
public class RendezvousHashCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(RendezvousHashCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static JsonElement Seed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "consistent-hash", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        return doc.RootElement.Clone();
    }

    private static ulong U(string s) => ulong.Parse(s, CultureInfo.InvariantCulture);

    [Fact]
    public void SeedsAgreeWithSeed()
    {
        var seeds = Seed().GetProperty("seeds");
        var expected = seeds.GetProperty("result").EnumerateArray().Select(e => U(e.GetString()!)).ToArray();
        Assert.Equal<IReadOnlyList<ulong>>(expected, RendezvousHash.Seeds(seeds.GetProperty("n").GetInt32()));
    }

    [Fact]
    public void PickAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("pick").EnumerateArray())
        {
            Assert.Equal(v.GetProperty("result").GetInt32(),
                RendezvousHash.Pick(v.GetProperty("buckets").GetInt32(), U(v.GetProperty("key").GetString()!)));
        }
    }
}
